### PR TITLE
AAP-65054 Fix bugs where concurrent jobs would clear facts of unrelated hosts

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
Describing key fixes:

 - in `finish_fact_cache`, we grabbed hosts by `name` only, yet, names can repeat in multiple inventories. Very surprising this wasn't caught before.
 - Again, in `finish_fact_cache`, if multiple conditions are true, we incorrectly conclude that "meta: clear_facts" was ran, which it wasn't actually. This results in clearing facts from a host not listed in `limit`. Those multiple conditions are:
   - Host did not have any facts with job began
   - Host had facts written by another job while the current job was in progress
   - The host was not matched by the current job's `limit`

So in the 2nd case here, the _absence_ of facts for the Host is taken as evidence that "clear_facts" happened, when in reality the host had no facts when the job started.

Why currently draft:
 - ~The tests take too long. They need to changed to run dramatically faster, or be removed.~
   - took the sleep down to 2 seconds, if this is insufficient the test will fail
 - ~Probably need some tests in `awx/main/tests/functional` to compliment or replace these~
   - added, a little slop-y but fine.
 - ~Need to clean up logging, some is kind of debugging sort of stuff.~
 - ~Unclear if this has more to think about with constructed inventory.~
   - developed a new fix & test for constructed inventory
 - [ABANDONED] I was going to not update facts if the 2-second scenario is suspected, but having the other fixes in, I can not force any actual test failure to happen. This might still be valid in the future, but not seeing it now so not doing anything.

##### ISSUE TYPE
 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME
 - API

---

> [!NOTE]
> **Medium Risk**
> Touches core fact-cache write/clear logic used by job runs, so mistakes could cause missed updates or stale facts, but changes are localized and covered by new functional/live tests.
> 
> **Overview**
> Fixes `finish_fact_cache` to **only update/clear facts for the intended hosts** by taking an explicit host QuerySet (scoped to the job’s inventory/slice) instead of querying by hostname globally.
> 
> Adds concurrency protection so a missing fact file will *not* clear a host if its `ansible_facts_modified` indicates another job updated facts after this job started, and improves reliability of detecting Ansible-updated fact files by backdating reference file mtimes.
> 
> Adds constructed-inventory handling (`Job.get_source_hosts_for_constructed_inventory`) so fact writes/clears target the *source* hosts, plus new functional/live/unit tests and a slow-gather playbook to reproduce concurrent `--limit` scenarios.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 9fb00a0eb79c1ccceb7fa38c79db39d5e63a4a34. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Fact cache now backdates written fact files, tracks per-host update/clear counts with batched processing, and scopes updates using an explicit host set (including correct handling for constructed inventories).

* **Bug Fixes**
  * Prevents clearing facts when a host was updated after a job started, ensures fact updates are limited to the intended hosts/inventory, and improves logging of update/clear totals.

* **Tests**
  * Added comprehensive unit, functional, and live tests covering concurrency, inventory scoping, constructed inventories, and edge cases.

### CODE REVIEW

File: awx/main/tasks/facts.py [hunk 1]
```diff
@@ -95,7 +95,7 @@ def finish_fact_cache(artifacts_dir, job_id=None, inventory_id=None, log_data=No
         return
 
     host_names = summary.get('hosts_cached', [])
-    hosts_cached = Host.objects.filter(name__in=host_names).order_by('id').iterator()
+    hosts_cached = Host.objects.filter(name__in=host_names, inventory_id=inventory_id).order_by('id').iterator()
```
Comment (AlanCoding) [inline awx/main/tasks/facts.py]: Seems to be introduced in https://github.com/ansible/awx/pull/15970, and this was a novel approach, so it may be that it doesn't quite work for constructed inventory.

**Comment by AlanCoding** [comment]
Tests locally run in:

> 4 passed, 2 warnings in 30.08s

This isn't too surprising, considering that the playbooks have a 2-second sleep in them... times 4 tests and we're in the order-of-magnitude here. This is a little heavy testing-wise, but I think we're getting a lot out of these.

**Review by djyasin — APPROVED** [review]

File: awx/main/tasks/facts.py [hunk 2]
```diff
@@ -110,6 +110,21 @@ def finish_fact_cache(artifacts_dir, job_id=None, inventory_id=None, log_data=No
...
             modified = os.path.getmtime(filepath)
             if not facts_write_time or modified >= facts_write_time:
+                # Zipfile (used for artifact transfer) truncates timestamps to 2-second
+                # resolution (DOS time format).  When the playbook writes facts very
+                # quickly after job start, the fact file and the summary file can be
+                # rounded to the same even second, making it impossible to tell whether
+                # the playbook actually touched the file or it is an unchanged reference.
+                if facts_write_time and (modified - facts_write_time) < 2 and modified % 1.0 < 1e-6 and facts_write_time % 1.0 < 1e-6:
+                    logger.warning(
+                        f'Ambiguous fact file timestamp for host {smart_str(host.name)} '
```
Comment (fosterseth) [inline awx/main/tasks/facts.py]: good thinking. if we ever run into issues where facts are not be updating when they should, we can check for this log. Then we could just add a time.sleep(2) or something after writing local facts files

File: awx/main/tasks/facts.py [hunk 3]
```diff
@@ -138,14 +153,25 @@ def finish_fact_cache(artifacts_dir, job_id=None, inventory_id=None, log_data=No
...
+                hosts_to_update.append(host)
+                logger.info(f'Facts cleared for inventory {smart_str(host.inventory.name)} host {smart_str(host.name)}')
+                log_data['cleared_ct'] += 1
 
         if len(hosts_to_update) >= 100:
             bulk_update_sorted_by_id(Host, hosts_to_update, fields=['ansible_facts', 'ansible_facts_modified'])
             hosts_to_update = []
 
     bulk_update_sorted_by_id(Host, hosts_to_update, fields=['ansible_facts', 'ansible_facts_modified'])
+    logger.debug(f'Updated {len(hosts_to_update)} host facts for inventory {inventory_id} in job {job_id}')
```
[the remark this answers is not in the document]
Comment (fosterseth) [reply awx/main/tasks/facts.py]: seems valid @AlanCoding

File: awx/main/tests/live/tests/test_concurrent_fact_cache.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+
+    wait_for_job(job_a)
+    wait_for_job(job_b)
+
+    host_0 = inv.hosts.get(name='cc_host_0')
+    host_1 = inv.hosts.get(name='cc_host_1')
+
+    # Both hosts should have the UPDATED value, not the old seed value
+    discovered_foos = [host_0.ansible_facts.get('foo'), host_1.ansible_facts.get('foo')]
+    assert all(discovered_foos), f'Facts were reverted to stale values by concurrent job cc_host_0 or _1: {discovered_foos}'
```
[the remark this answers is not in the document]
Comment (fosterseth) [reply awx/main/tests/live/tests/test_concurrent_fact_cache.py]: also seems valid

**Review by fosterseth — APPROVED** [review]
i'm good with this once the bot's comments are addressed

File: awx/main/tasks/facts.py (hunk 2 above)
Comment (AlanCoding) [reply awx/main/tasks/facts.py]: But that would cost 2 seconds!

Seriously though, I do feel an itch to employ a method of "backdating" the files laid down. So get the modtime of the files, subtract 2, and then change the modtime to that value. That should address it without sacrificing 2 seconds. But right now testing is not showing this issue. Perhaps if we make the job run code more efficient we might risk seeing that.

File: awx/main/tasks/facts.py [hunk 5]
```diff
@@ -110,6 +110,21 @@ def finish_fact_cache(artifacts_dir, job_id=None, inventory_id=None, log_data=No
...
+                if facts_write_time and (modified - facts_write_time) < 2 and modified % 1.0 < 1e-6 and facts_write_time % 1.0 < 1e-6:
+                    logger.warning(
+                        f'Ambiguous fact file timestamp for host {smart_str(host.name)} '
+                        f'in job {job_id} inventory {inventory_id}: '
+                        f'fact_mtime={modified}, summary_mtime={facts_write_time}, '
+                        f'delta={modified - facts_write_time:.1f}s. '
+                        f'Both timestamps lack fractional seconds, consistent with '
+                        f'zipfile 2-second rounding. Cannot reliably determine if '
+                        f'the file was modified by the playbook.'
+                    )
```
[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/tasks/facts.py]: Not doing this until I can get a test reproducer.

It's also fairly unrealistic for any real customer circumstance. I want to see logs to justify the action before we do anything like that.

[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/tasks/facts.py]: Ok I was wrong, we are hitting this.

```
tools_awx_1       | 2026-03-03 20:18:33,964 WARNING  [-] awx.main.tasks.facts Ambiguous fact file timestamp for host cc_host_0 in job 303 inventory 159: fact_mtime=1772569110.0, summary_mtime=1772569110.0, delta=0.0s. Both timestamps lack fractional seconds, consistent with zipfile 2-second rounding. Cannot reliably determine if the file was modified by the playbook.
tools_awx_1       | 2026-03-03 20:18:33,964 INFO     [-] awx.main.tasks.facts New fact for inventory test_concurrent_fact_cache host cc_host_0
tools_awx_1       | 2026-03-03 20:18:33,964 WARNING  [-] awx.main.tasks.facts Ambiguous fact file timestamp for host cc_host_1 in job 303 inventory 159: fact_mtime=1772569110.0, summary_mtime=1772569110.0, delta=0.0s. Both timestamps lack fractional seconds, consistent with zipfile 2-second rounding. Cannot reliably determine if the file was modified by the playbook.
tools_awx_1       | 2026-03-03 20:18:33,964 INFO     [-] awx.main.tasks.facts New fact for inventory test_concurrent_fact_cache host cc_host_1
```

And improving the assertions make it clear...

[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/tasks/facts.py]: Ok I hate that the bot was right. But I had to go ahead and implement the backdating solution. I think this is still fairly strong and the most correct we can get.

File: awx/main/tests/live/tests/test_concurrent_fact_cache.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/tests/live/tests/test_concurrent_fact_cache.py]: This was a bit of a pita, and required some real non-trivial code changes.

File: awx/main/tasks/facts.py [hunk 6]
```diff
@@ -138,14 +145,25 @@ def finish_fact_cache(artifacts_dir, job_id=None, inventory_id=None, log_data=No
...
+                    f'created time ({job_created.isoformat()}). '
+                    f'A concurrent job likely updated this host\'s facts while this job was running.'
+                )
+                log_data['unmodified_ct'] += 1
+            else:
+                host.ansible_facts = {}
+                host.ansible_facts_modified = now()
+                hosts_to_update.append(host)
+                logger.info(f'Facts cleared for inventory {smart_str(host.inventory.name)} host {smart_str(host.name)}')
+                log_data['cleared_ct'] += 1
```
[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/tasks/facts.py:163]: > A concurrent job can update the same host in between and still get overwritten here.

Yeah, but in that case it's unclear which job _should_ save their variables!

The fact clearing bug is a true misinterpretation. Concurrent jobs running in the same time window and both saving facts is just thrashing due... to the nature of concurrent jobs. The one that finishes last should write, which is effectively the behavior we already get, and this is not changed.

File: awx/main/tasks/facts.py (hunk 2 above)
Comment (AlanCoding) [reply awx/main/tasks/facts.py]: didn't originally mean to, but hitting more test failures, I found that I needed to implement backdating to get correct behavior, and I'm struggling with a solution that does anything less.

File: awx/main/tasks/jobs.py [hunk 7]
```diff
@@ -1224,10 +1224,16 @@ def post_run_hook(self, job, status):
             return
         if self.should_use_fact_cache() and self.runner_callback.artifacts_processed:
             job.log_lifecycle("finish_job_fact_cache")
+            if job.inventory.kind == 'constructed':
+                hosts_qs = job.get_hosts_for_fact_cache()
```
[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/tasks/jobs.py]: Not true, because this won't happen in the constructed inventory case.

But that's not obviously from the implied types, so I made a code change to address this.

File: awx/main/models/jobs.py [hunk 8]
```diff
@@ -845,24 +845,37 @@ def get_notification_templates(self):
...
+        host in the input inventory.  This resolves those references and returns
+        a proper QuerySet (never a list), suitable for use with finish_fact_cache.
+        """
+        Host = JobHostSummary._meta.get_field('host').related_model
+        if not self.inventory_id:
+            return Host.objects.none()
+        id_field = Host._meta.get_field('id')
+        return Host.objects.filter(id__in=self.inventory.hosts.exclude(instance_id='').values_list(Cast('instance_id', output_field=id_field))).only(
+            *HOST_FACTS_FIELDS
+        )
```
[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/models/jobs.py:861]: This hasn't changed from the prior solution. That relies on constructed inventory only using this type of `instance_id` from the get-go.

Although I would be open to a fix, `.filter(instance_id__regex=r'^\d+$')` is too complicating of the issued SQL

File: awx/main/tasks/jobs.py [hunk 9]
```diff
@@ -1224,10 +1224,16 @@ def post_run_hook(self, job, status):
             return
         if self.should_use_fact_cache() and self.runner_callback.artifacts_processed:
             job.log_lifecycle("finish_job_fact_cache")
+            if job.inventory.kind == 'constructed':
+                hosts_qs = job.get_source_hosts_for_constructed_inventory()
+            else:
+                hosts_qs = job.inventory.hosts
```
[the remark this answers is not in the document]
Comment (AlanCoding) [reply awx/main/tasks/jobs.py:1308]: We were not previously using `.only(*HOST_FACTS_FIELDS)` in finish_fact_cache.

While this comment isn't totally invalid, it's not a big thing on my mind.
